### PR TITLE
introduce ovirt-external-provider role

### PR DESCRIPTION
**Pull request by tareqalayan**

### CODE REVIEW

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Review by mwperina — COMMENTED** [review]
Why do we need a role that when we only call the module and not using this role inside other roles? Am I missing something or are you planning to introduce using this role inside other roles in upcoming PRs?

**Comment by tareqalayan** [comment]
@mwperina I was intending to add this role to be part of the ovirt-infra role.

**Comment by mwperina** [comment]
@tareqalayan So could you please add it to ovirt-infra within this PRs so it's clear why this role is added?

**Comment by mwperina** [comment]
ci test please

File: roles/ovirt-external-providers/.travis.yml [hunk 1]
```diff
@@ -0,0 +1,29 @@
+---
```
Comment (machacekondra) [inline roles/ovirt-external-providers/.travis.yml]: please remove this file

File: roles/ovirt-external-providers/README.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
+# TODO
```
Comment (machacekondra) [inline roles/ovirt-external-providers/README.md]: Why todo?

File: roles/ovirt-external-providers/README.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+oVirt External Providers
+========================
+
+The `ovirt-external-providers` role is used set up oVirt external providers like glance.
```
Comment (machacekondra) [inline roles/ovirt-external-providers/README.md]: I would remove glance. Not everybody knows what that is.

File: roles/ovirt-external-providers/README.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+|-----------------------|-----------------------|-----------------------------------------|
+| external_providers    | UNDEF                 | List of dictionaries that describe the external provider. |
+
+The `external_providers` list can contain the following parameters:
+
+| Name                              | Default value       | Description                             |
+|-----------------------------------|---------------------|-----------------------------------------|
+| name                              | UNDEF               | Name of the external provider.                     |
+| state                             | present             | State of the external provider.                    |
+| type                              | UNDEF               | Type of the external provider. Values can be os_image, network, os_volume and  foreman. |
```
Comment (machacekondra) [inline roles/ovirt-external-providers/README.md]: Can you please format it as follows::

```
Type of the external provider, values can be:
<ul>
<li>os_image</li>
<li>os_volume</li>
...
</ul>
```

File: roles/ovirt-external-providers/README.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+| url                               | UNDEF               | URL where external provider is hosted. |
+| username                          | UNDEF               | Username to be used for login to external provider. Applicable for all types. |
+| password                          | UNDEF               | Password of the user specified in username parameter. Applicable for all types. |
+| tenant                            | UNDEF               | Name of the tenant. |
+| auth_url                          | UNDEF               | Keystone authentication URL of the openstack provider. required: os_image, os_volume and network. |
```
Comment (machacekondra) [inline roles/ovirt-external-providers/README.md]:
```
Keystone authentication URL of the openstack provider. Required for following external provider types
<ul>
<li>...</li>
...
</ul>
```

File: roles/ovirt-external-providers/README.md [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+----------------
+
+```yaml
+- name: oVirt infra
+  hosts: localhost
+  connection: local
+  gather_facts: false
+
+  vars:
+   data_center_name: mydatacenter
```
Comment (machacekondra) [inline roles/ovirt-external-providers/README.md]: Why the `dc` vars are needed?

File: roles/ovirt-external-providers/defaults/main.yml [hunk 7]
```diff
@@ -0,0 +1,2 @@
+---
```
Comment (machacekondra) [inline roles/ovirt-external-providers/defaults/main.yml]: can you please remove this file

File: roles/ovirt-external-providers/handlers/main.yml [hunk 8]
```diff
@@ -0,0 +1,2 @@
+---
```
Comment (machacekondra) [inline roles/ovirt-external-providers/handlers/main.yml]: can you please remove this file

File: roles/ovirt-external-providers/tests/inventory [hunk 9]
```diff
@@ -0,0 +1 @@
+localhost
```
Comment (machacekondra) [inline roles/ovirt-external-providers/tests/inventory]: can you please remove this file

File: roles/ovirt-external-providers/tests/test.yml [hunk 10]
```diff
@@ -0,0 +1,5 @@
+---
```
Comment (machacekondra) [inline roles/ovirt-external-providers/tests/test.yml]: can you please remove this file

File: roles/ovirt-external-providers/vars/main.yml [hunk 11]
```diff
@@ -0,0 +1,2 @@
+---
```
Comment (machacekondra) [inline roles/ovirt-external-providers/vars/main.yml]: can you please remove this file

File: roles/ovirt-external-providers/README.md (hunk 5 above)
Comment (machacekondra) [inline roles/ovirt-external-providers/README.md]: I miss the `data center` variable here.

File: roles/ovirt-infra/meta/main.yml [hunk 12]
```diff
@@ -26,3 +26,4 @@ dependencies:
   - ovirt-storages
   - ovirt-aaa-jdbc
   - ovirt-permissions
+  - ovirt-external-providers
```
Comment (machacekondra) [inline roles/ovirt-infra/meta/main.yml]: Maybe we should call it before ovirt-permission, so maybe someone want to setup permissions for the external provider...

**Review by machacekondra — CHANGES_REQUESTED** [review]

File: roles/ovirt-external-providers/.travis.yml (hunk 1 above)
Comment (tareqalayan) [reply roles/ovirt-external-providers/.travis.yml]: OK

File: roles/ovirt-external-providers/README.md (hunk 2 above)
Comment (tareqalayan) [reply roles/ovirt-external-providers/README.md]: done

File: roles/ovirt-external-providers/README.md (hunk 3 above)
Comment (tareqalayan) [reply roles/ovirt-external-providers/README.md]: done

File: roles/ovirt-external-providers/README.md (hunk 6 above)
Comment (tareqalayan) [reply roles/ovirt-external-providers/README.md]: not needed

File: roles/ovirt-external-providers/README.md (hunk 5 above)
Comment (tareqalayan) [reply roles/ovirt-external-providers/README.md]: not needed

File: roles/ovirt-external-providers/README.md (hunk 6 above)
Comment (tareqalayan) [reply roles/ovirt-external-providers/README.md]: done

File: roles/ovirt-external-providers/defaults/main.yml (hunk 7 above)
Comment (tareqalayan) [reply roles/ovirt-external-providers/defaults/main.yml]: done

File: roles/ovirt-external-providers/handlers/main.yml (hunk 8 above)
Comment (tareqalayan) [reply roles/ovirt-external-providers/handlers/main.yml]: done

File: roles/ovirt-external-providers/tests/inventory (hunk 9 above)
Comment (tareqalayan) [reply roles/ovirt-external-providers/tests/inventory]: done

File: roles/ovirt-external-providers/tests/test.yml (hunk 10 above)
Comment (tareqalayan) [reply roles/ovirt-external-providers/tests/test.yml]: done

File: roles/ovirt-external-providers/vars/main.yml (hunk 11 above)
Comment (tareqalayan) [reply roles/ovirt-external-providers/vars/main.yml]: done

File: roles/ovirt-infra/meta/main.yml (hunk 12 above)
Comment (tareqalayan) [reply roles/ovirt-infra/meta/main.yml]: done

**Review by tareqalayan — COMMENTED** [review]
fixed according to comment

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

**Comment by ovirt-infra** [comment]
The commiter is not a member of organization nor whitelist. Please contact maintainer for patch verification.

File: roles/ovirt-external-providers/README.md (hunk 5 above)
Comment (machacekondra) [reply roles/ovirt-external-providers/README.md]: So why do you pass it in?

 - roles/ovirt-external-providers/tasks/main.yml

Either remove it or document it.

**Review by machacekondra — CHANGES_REQUESTED** [review]
It's improperly formatted, please verify the format of markdown.

File: roles/ovirt-external-providers/tasks/main.yml [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+  ovirt_external_providers:
+    auth: "{{ ovirt_auth }}"
+    name: "{{ item.name }}"
+    state: "{{ item.state }}"
+    type: "{{ item.type }}"
+    url: "{{ item.url }}"
+    username: "{{ item.username }}"
+    password: "{{ item.password }}"
+    tenant: "{{ item.tenant }}"
+    auth_url: "{{ item.auth_url }}"
```
Comment (machacekondra) [inline roles/ovirt-external-providers/tasks/main.yml]: I think most of those paramaters are not mandatory, so we should use `default(omit)`

File: roles/ovirt-infra/README.md [hunk 14]
```diff
@@ -51,6 +51,10 @@ The role has dependencies on multiple other oVirt roles. The following list disp
 
  This role accepts a variable called `permissions`. 
 
+From [ovirt-external-providers]
+
+This role expects a list called `external_providers`, which contains information about the external provider. To see example of this list please check the role's documentation.
```
Comment (machacekondra) [inline roles/ovirt-infra/README.md]: Please change this to:
```
This role accepts a variable called `external_providers`. 
```

It's new documentation format.

File: roles/ovirt-external-providers/tasks/main.yml [hunk 15]
```diff
@@ -0,0 +1,17 @@
+---
+- name: Add external providers
+  ovirt_external_providers:
+    auth: "{{ ovirt_auth }}"
+    name: "{{ item.name }}"
+    state: "{{ item.state }}"
```
Comment (machacekondra) [inline roles/ovirt-external-providers/tasks/main.yml]: In doc you say the `default` is `present` which isn't here.

Comment (tareqalayan) [reply roles/ovirt-external-providers/tasks/main.yml]: done

File: roles/ovirt-infra/README.md (hunk 14 above)
Comment (tareqalayan) [reply roles/ovirt-infra/README.md]: done

File: roles/ovirt-external-providers/tasks/main.yml (hunk 13 above)
Comment (machacekondra) [reply roles/ovirt-external-providers/tasks/main.yml]: Does all providers have mandatory ^those attributes?

Comment (tareqalayan) [reply roles/ovirt-external-providers/tasks/main.yml]: Yes i need to pass 
    name: "{{ item.name }}"
    state: "{{ item.state | default('present') }}"
    type: "{{ item.type }}"
    url: "{{ item.url }}"
    username: "{{ item.username  }}"
    password: "{{ item.password  }}"
    tenant: "{{ item.tenant   }}"
    auth_url: "{{ item.auth_url }}"
    data_center: "{{ item.data_center | default(omit) }}"
 
for the ovirt public glance i passed dummy data on those parameters and it added it. 

how to proceed from here?

File: roles/ovirt-infra/README.md [hunk 16]
```diff
@@ -51,6 +51,10 @@ The role has dependencies on multiple other oVirt roles. The following list disp
 
  This role accepts a variable called `permissions`. 
 
+From [ovirt-external-providers]
+
+This role accepts a list called `external_providers`, which contains information about the external provider. To see example of this list please check the role's documentation.
```
Comment (machacekondra) [inline roles/ovirt-infra/README.md]: Can you change to:
```
This role accepts a variable called `external_providers`.
```
As we have in all other roles?

**Review by machacekondra — APPROVED** [review]
Looks good , thanks, one last comment and I will merge

File: roles/ovirt-infra/README.md (hunk 16 above)
Comment (tareqalayan) [reply roles/ovirt-infra/README.md]: OK

File: roles/ovirt-infra/README.md [hunk 17]
```diff
@@ -51,6 +51,11 @@ The role has dependencies on multiple other oVirt roles. The following list disp
 
  This role accepts a variable called `permissions`. 
 
+From [ovirt-external-providers]
+
+This role accepts a variable called `external_providers`, which contains information about the external provider.
+To see example of this list please check the role's documentation.
```
Comment (machacekondra) [inline roles/ovirt-infra/README.md]: I don't see it's changed?

Comment (tareqalayan) [reply roles/ovirt-infra/README.md]: ok

**Pull request merged**